### PR TITLE
chore(core): fix WAL append crash for varchar column

### PR DESCRIPTION
**Pull request by bluestreak01**
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f0a2aed15ef, pid=1934, tid=3258
#
# JRE version: OpenJDK Runtime Environment Temurin-11.0.22+7 (11.0.22+7) (build 11.0.22+7)
# Java VM: OpenJDK 64-Bit Server VM Temurin-11.0.22+7 (11.0.22+7, mixed mode, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# C  [libquestdb2941880153616460049.so+0x7e5ef]  set_varchar_null_refs_AVX2(long*, long, long)+0x5f
#
# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport -p%p -s%s -c%c -d%d -P%P -u%u -g%g -- %E" (or dumping to /home/vsts/work/1/s/core/core.1934)
#
# If you would like to submit a bug report, please visit:
#   https://github.com/adoptium/adoptium-support/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

---------------  S U M M A R Y ------------

Command Line: -ea -Dfile.encoding=UTF-8 /home/vsts/work/1/s/core/target/surefire/surefirebooter6109048573964084745.jar /home/vsts/work/1/s/core/target/surefire/surefire11576409589169175072tmp /home/vsts/work/1/s/core/target/surefire/surefire_07837652167085883151tmp

Host: Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz, 2 cores, 6G, Ubuntu 22.04.4 LTS
Time: Mon Apr  1 01:58:41 2024 UTC elapsed time: 374.041564 seconds (0d 0h 6m 14s)

---------------  T H R E A D  ---------------

Current thread (0x00007f09d83d2800):  JavaThread "Thread-705" daemon [_thread_in_native, id=3258, stack(0x00007f09c8bfd000,0x00007f09c8cfd000)]

Stack: [0x00007f09c8bfd000,0x00007f09c8cfd000],  sp=0x00007f09c8cfaed8,  free space=1015k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libquestdb2941880153616460049.so+0x7e5ef]  set_varchar_null_refs_AVX2(long*, long, long)+0x5f
j  io.questdb.cairo.VarcharTypeDriver.setColumnRefs(JJJ)V+4
j  io.questdb.cairo.wal.WalWriter.setVarColumnFixedFileNull(Lio/questdb/cairo/ColumnTypeDriver;IJI)V+58
J 7850 c1 io.questdb.cairo.wal.WalWriter.setColumnNull(IIJI)V (44 bytes) @ 0x00007f0a0dc65834 [0x00007f0a0dc65600+0x0000000000000234]
J 4601 c1 io.questdb.cairo.wal.WalWriter$MetadataWriterService.addColumn(Ljava/lang/CharSequence;IIZZIZLio/questdb/cairo/SecurityContext;)V (515 bytes) @ 0x00007f0a0d44c68c [0x00007f0a0d44b160+0x000000000000152c]
J 4356 c1 io.questdb.griffin.engine.ops.AlterOperation.applyAddColumn(Lio/questdb/cairo/wal/MetadataService;)V (226 bytes) @ 0x00007f0a0d3cbddc [0x00007f0a0d3cb580+0x000000000000085c]
J 4355 c1 io.questdb.griffin.engine.ops.AlterOperation.apply(Lio/questdb/cairo/wal/MetadataService;Z)J (504 bytes) @ 0x00007f0a0d3c3fd4 [0x00007f0a0d3c2a40+0x0000000000001594]
J 7934 c1 io.questdb.cairo.wal.WalWriter.applyMetadataChangeLog(J)V (147 bytes) @ 0x00007f0a0dc92594 [0x00007f0a0dc921a0+0x00000000000003f4]
J 3482 c1 io.questdb.cairo.wal.WalWriter.getSequencerTxn()J (75 bytes) @ 0x00007f0a0d14edf4 [0x00007f0a0d14ed60+0x0000000000000094]
J 3480 c1 io.questdb.cairo.wal.WalWriter.commit()J (293 bytes) @ 0x00007f0a0d14cea4 [0x00007f0a0d14cd80+0x0000000000000124]
J 10513 c2 io.questdb.test.griffin.wal.FuzzRunner.lambda$createWalWriteThread$5(Ljava/util/concurrent/atomic/AtomicInteger;Lio/questdb/std/ObjList;Ljava/util/concurrent/ConcurrentLinkedQueue;Ljava/util/concurrent/atomic/AtomicLong;Ljava/util/concurrent/atomic/AtomicLong;Lio/questdb/std/ObjList;ILjava/lang/String;Ljava/util/concurrent/atomic/AtomicInteger;)V (450 bytes) @ 0x00007f0a14378f3c [0x00007f0a14378240+0x0000000000000cfc]
j  io.questdb.test.griffin.wal.FuzzRunner$$Lambda$526.run()V+40
J 8027 c1 java.lang.Thread.run()V java.base@11.0.22 (17 bytes) @ 0x00007f0a0cdf1474 [0x00007f0a0cdf1340+0x0000000000000134]
v  ~StubRoutines::call_stub
V  [libjvm.so+0x8dfedb]  JavaCalls::call_helper(JavaValue*, methodHandle const&, JavaCallArguments*, Thread*)+0x39b
V  [libjvm.so+0x8dde9d]  JavaCalls::call_virtual(JavaValue*, Handle, Klass*, Symbol*, Symbol*, Thread*)+0x1ed
V  [libjvm.so+0x98c6ec]  thread_entry(JavaThread*, Thread*)+0x6c
V  [libjvm.so+0xee06da]  JavaThread::thread_main_inner()+0x1ba
V  [libjvm.so+0xedd31f]  Thread::call_run()+0x14f
V  [libjvm.so+0xc7b6b6]  thread_native_entry(Thread*)+0xe6
```

### CODE REVIEW

File: core/src/main/java/io/questdb/cairo/StringTypeDriver.java [hunk 1]
```diff
@@ -287,7 +290,7 @@ public long setAppendPosition(long pos, MemoryMA auxMem, MemoryMA dataMem) {
 
     @Override
     public void setColumnRefs(long address, long initialOffset, long count) {
-        Vect.setVarColumnRefs32Bit(address, initialOffset, count);
+        Vect.setVarColumnRefs32Bit(address, initialOffset, count + 1);
```
Comment (jerrinot) [inline core/src/main/java/io/questdb/cairo/StringTypeDriver.java:293]: this is a behaviour change in all other places except `WalWriter` (where you compensated for this +1)

for example here: https://github.com/jerrinot/questdb/blob/e065400768734bde734149b1272b0f2f8968f779/core/src/main/java/io/questdb/cairo/O3OpenColumnJob.java#L336

are you sure we are not trading one segfault for another? (I don't understand `O3OpenColumnJob.o3PartitionMerge()` well enough to tell without investigating how it works)

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 11 / 11 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/VarcharTypeDriver.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/BinaryTypeDriver.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/StringTypeDriver.java|4|4|100.00%|

File: core/src/main/java/io/questdb/cairo/StringTypeDriver.java (hunk 1 above)
Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/StringTypeDriver.java:293]: we won't have another seg fault fortunately, but driver API is vague, i agree. It has to be more explicit about the intent.

Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/StringTypeDriver.java]: trying to fix here: https://github.com/questdb/questdb/pull/4362/commits/a9fea2bb81d19ab7f88efabafe82afe10d61792b

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 31 / 32 (96.88%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|10|11|90.91%|
|:large_blue_circle:|io/questdb/cairo/frm/file/ContiguousFileVarFrameColumn.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/VarcharTypeDriver.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3OpenColumnJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/BinaryTypeDriver.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/StringTypeDriver.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 31 / 32 (96.88%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|10|11|90.91%|
|:large_blue_circle:|io/questdb/cairo/frm/file/ContiguousFileVarFrameColumn.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/VarcharTypeDriver.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3OpenColumnJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/BinaryTypeDriver.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/StringTypeDriver.java|7|7|100.00%|

**Review by jerrinot — APPROVED** [review]

**Pull request merged**